ord_schema/validations.py

```diff
@@ -120,16 +127,78 @@ def validate_reaction_identifier(message):
     if len(message.components) == 0:
         raise ValueError('Reaction inputs must have at least one component')
+    for component in message.components:
+        if not component.WhichOneof('amount'):
+            raise ValueError('Reaction input\'s components require an amount')
     return message
 
 
+def pubchem_resolve(value_type, value):
+    return request.urlopen('https://pubchem.ncbi.nlm.nih.gov/rest/pug'
+                           f'/compound/{value_type}/{value}/property/'
+                           'IsomericSMILES/txt').read().decode().strip()
+
 def validate_compound(message):
+    # pylint: disable=too-many-branches
     if len(message.identifiers) == 0:
         raise ValueError('Compounds must have at least one identifier')
+
+    # If no structural identifiers have been used, iterate through available
+    # identifiers until we can resolve one to a SMILES string
+    if not any(identifier.type in COMPOUND_STRUCTURAL_IDENTIFIERS_ for
+               identifier in message.identifiers):
+        for identifier in message.identifiers:
+            if identifier.type == identifier.NAME:
+                try:
+                    smiles = pubchem_resolve('name', identifier.value)
+                    new_identifier = message.identifiers.add()
+                    new_identifier.type = new_identifier.SMILES
+                    new_identifier.value = smiles
+                    new_identifier.details = 'NAME resolved by PubChem'
+                    break
+                except HTTPError:
+                    pass
+
+    # Try to create an RDKit binary identifier
+    # TODO(ccoley) add more sources for RDKIT_BINARY
+    if Chem and not any(identifier.type == identifier.RDKIT_BINARY for
+                        identifier in message.identifiers):
+        mol = None
+        for identifier in message.identifiers:
+            this_mol = None
+
+            if Chem and identifier.type == identifier.SMILES:
+                this_mol = Chem.MolFromSmiles(identifier.value)
+                if this_mol is None:
+                    raise ValueError(f'RDKit {RDKIT_VERSION} could not validate'
+                                     f' SMILES identifier {identifier.value}')
+            elif identifier.type == identifier.INCHI:
+                this_mol = Chem.MolFromInchi(identifier.value)
+                if this_mol is None:
+                    raise ValueError(f'RDKit {RDKIT_VERSION} could not validate'
+                                     f' InChI identifier {identifier.value}')
+            elif identifier.type == identifier.MOLBLOCK:
+                this_mol = Chem.MolFromMolBlock(identifier.value)
+                if this_mol is None:
+                    raise ValueError(f'RDKit {RDKIT_VERSION} could not validate'
+                                     ' MolBlock identifier')
+
+            # If RDKit mol was not defined, define it with this resolved mol
+            mol = mol or this_mol
```

Comment:
I'm a bit confused by this; why not just `break` after getting any valid molecule and avoid the `this_mol` temporary variable?

Reply:
If a message has defined a SMILES and an InChI, we will want to validate both despite only needing one to define the ```RDKIT_BINARY``` identifier. Consider a case where the SMILES is valid but the InChI is valid; we don't want to break after resolving the valid SIMLES. We could even consider adding greater consistency checks by requiring all resolved molecules to be identical, but that is risky given the challenges of molecule canonicalization (charges, tautomers, etc.)

Reply:
SGTM; possibly add a comment to clarify this.